490 create sso UI

qhantom opened the pull request:
## DESCRIPTION

This PR adds support for single-sign-on (SSO) via any provider. This not only includes the ability to login but also to hide UI elements which are not relevant for the user:

- change mail address, field is disabled if `isSso` is `true`
- change password tab on profile page is hidden if `isSso` is `true` since source-of-truth are the OAuth provider

Beside those changes, I made further relevant changes for the SSO support:

- accept loading remote images from all frachtwerk.de addresses
- redirect user to backend endpoint if this pattern is satisfied: `'/oauth2/:path*'`
- refactored the architecture of the `Login` folder inside `lib` to make it possible to add SSO due to flexibility
- add a solution for `useGetRoles` to not only accept request props but also query props like `enabled` to manually disable a query
### [ℹ️ Watch Loomi Loom](https://www.loom.com/share/edda48473e3642cdb618ffb99c3cbf84)

### How to test SSO

Since SSO is depended on how the backend is configured, here is a step-by-step guide to test it locally with an experimental backend instance:

1. inside the `.env` file, replace the value at `NEXT_PUBLIC_API_BASE_URL` with the following: `https://essencium-experimental.staging.dev.frachtwerk.de` and the value at `NEXT_PUBLIC_API_URL` with this: `${NEXT_PUBLIC_API_BASE_URL}/${NEXT_PUBLIC_API_VERSION}`
2. spin up your Next.js dev server
3. you should see the GitLab SSO option on the login screen, choose this
4. authorise via GitLab
5. see redirect to home page of Essencium (since you have the `USER` role you cannot do much)
6. go to your profile and verify if you cannot change mail and password

### TO-DO

- [x] implement and update tests
- [x] update docs
- [x] pull `main` & resolve merge conflicts
- [x] check Vercel deployment

### CLOSES

Closes #490

PhilKsr reviewed (changes requested):
I guess there is no other user registered at `https://essencium-experimental.staging.dev.frachtwerk.de`, but I think we should test the combination of both login ways. Like log in with SSO, test all restrictions, log out - log in with normal admin user and check for any bugs. Should work in general but its safer to test. What do you think?

The loom and the migration guide was very helpful for the review! Such a good new feature and also good refactoring in the profile/roles! Thanks for the work.

qhantom commented:
> I guess there is no other user registered at `https://essencium-experimental.staging.dev.frachtwerk.de`, but I think we should test the combination of both login ways. Like log in with SSO, test all restrictions, log out - log in with normal admin user and check for any bugs. Should work in general but its safer to test. What do you think?
> 
> The loom and the migration guide was very helpful for the review! Such a good new feature and also good refactoring in the profile/roles! Thanks for the work.

Thanks! 🙏 There should be the default user, could you please check it again? I tested it with both login strategies and it worked fine.

qhantom commented on packages/app/src/pages/login/index.tsx:
```diff
@@ -70,18 +110,65 @@ function LoginView(): JSX.Element {
+                <Box mt="md">
+                  {Object.keys(ssoApplications).map(application => (
+                    <NextLink
+                      style={{ textDecoration: 'none', color: 'white' }}
```
Good point, I change it!

qhantom commented on packages/app/MIGRATION.md:
```diff
@@ -1,61 +1,408 @@
 - [Essencium migration PR](https://github.com/Frachtwerk/essencium-frontend/pull/496/files#top)
 - [Mantine v7](https://mantine.dev/changelog/7-0-0/)
 
-### `packages/app/postcss.config.cjs`
```
What do you mean here?

PhilKsr reviewed (changes requested):
Looks good with the admin user! I just ran into an edgecase, I think we need a few small additional changes:
https://www.loom.com/share/60e6ae389b144d52a01d2455ab49bd5d?sid=c84915b6-cf1e-4efc-abff-a5009a78618e

qhantom commented:
> Looks good with the admin user! I just ran into an edgecase, I think we need a few small additional changes: https://www.loom.com/share/60e6ae389b144d52a01d2455ab49bd5d?sid=c84915b6-cf1e-4efc-abff-a5009a78618e

Very important take from you, thanks @PhilKsr, I'm on it 🚀

qhantom commented:
@PhilKsr I pushed new commits to this PR:
- all files I touched have been refactored concerning CSS modules migration
- the backend provides a `source` property on the user object which I utilise to check for SSO login
- update test

I [recorded another Loom](https://www.loom.com/share/9c226611bf0a420280b1ae2eb6725fed) for better understanding

qhantom commented on packages/app/src/api/auth.ts:
```diff
@@ -152,3 +162,29 @@ export function useSetPassword(): UseMutationResult<
 }
+
+export const isSsoAtom = atom(get =>
+  Boolean(get(userAtom)?.source !== UserSource.LOCAL),
```
Nope, can be removed, thanks!

qhantom commented on packages/app/src/pages/login/index.tsx:
```diff
@@ -70,18 +108,69 @@ function LoginView(): JSX.Element {
-    />
+    <Container mt="150px">
+      {!oauthToken ? (
+        <Card shadow="sm" radius="sm" className={classes.loginCard} withBorder>
+          <Flex direction="column">
+            <Title order={2} className={classes['loginCard--title']}>
```
Yes good point, I think Prettier automatically formatted it.

qhantom commented on packages/lib/src/components/User/UserForm.tsx:
```diff
@@ -322,13 +344,14 @@ export function UserForm({
             render={({ field }) => (
               <MultiSelect
                 {...field}
+                disabled={isSso}
```
No but the roles are managed by the SSO provider like GitLab and not the backend 🙂

qhantom commented:
> Thanks for the work again! There are some small questions on my side.
> 
> In addition I think the Sso label in the header is a bit too much for the small box on the right. Maybe it would be an option to add it to the user profile instead (left side box for example). What do you think? <img alt="Screenshot 2024-03-12 at 09 46 53" width="1658" src="https://private-user-images.githubusercontent.com/93935879/312003907-b461676d-cc76-47d4-b043-c42a76fe38dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAyMzk5MjUsIm5iZiI6MTcxMDIzOTYyNSwicGF0aCI6Ii85MzkzNTg3OS8zMTIwMDM5MDctYjQ2MTY3NmQtY2M3Ni00N2Q0LWIwNDMtYzQyYTc2ZmUzOGRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEyVDEwMzM0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1MmE2MTIyYTE2N2ZlYjVhOWM2ZDg0YWZhZDdkYmEzMTZmMzBmMjc3ZDk3MzRmOTQ3M2FjMDYyNDRjOTgwY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8jLhBQ882rnHSZtEJ8oeADns8u6r_IyzxbztsE1Bpv4">

The intention was do clearly see with one blink with which user I'm currently logged in. When it's inside the profile card, it's one click away you need to know. For me, I think the spacing is fine. What do you think @CathrinTruchan?

CathrinTruchan commented:
> > Thanks for the work again! There are some small questions on my side.
> > In addition I think the Sso label in the header is a bit too much for the small box on the right. Maybe it would be an option to add it to the user profile instead (left side box for example). What do you think? <img alt="Screenshot 2024-03-12 at 09 46 53" width="1658" src="https://private-user-images.githubusercontent.com/93935879/312003907-b461676d-cc76-47d4-b043-c42a76fe38dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAyMzk5MjUsIm5iZiI6MTcxMDIzOTYyNSwicGF0aCI6Ii85MzkzNTg3OS8zMTIwMDM5MDctYjQ2MTY3NmQtY2M3Ni00N2Q0LWIwNDMtYzQyYTc2ZmUzOGRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEyVDEwMzM0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1MmE2MTIyYTE2N2ZlYjVhOWM2ZDg0YWZhZDdkYmEzMTZmMzBmMjc3ZDk3MzRmOTQ3M2FjMDYyNDRjOTgwY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8jLhBQ882rnHSZtEJ8oeADns8u6r_IyzxbztsE1Bpv4">
> 
> The intention was to clearly see with one blink with which user I'm currently logged in. When it's inside the profile card, it's one click away you need to know. For me, I think the spacing is fine. What do you think @CathrinTruchan?
> 
> UPDATE: I agree with @PhilKsr that the information for the user isn't that necessary and it's better placed inside the profile card. But lets wait for @CathrinTruchan opinion 🙂

I agree, the profile card is a good place for this information, since it is not super important for the user 🙂

qhantom commented:
@PhilKsr I removed the SSO badge and put it into the profile. I realised, that the badge looks like the active-badge of the user so I refactored it into an indicator and put it on the avatar. Please let me know what you think. I also added further tests for the profile components. You can give it another review 🚀

PhilKsr reviewed (approved):
Works perfectly and I really like the new active dot!
